### PR TITLE
[Heatmap Filtering] Make top taxa per sample a client-side filter

### PR DESCRIPTION
**Pull request by j-x-han**
# Description

Implement the Taxa per Sample slider on the client rather than making a server call for users with the `heatmap_filter_fe` flag enabled (see IDSEQ-2186).

# Notes

This PR doesn't include any backend changes to load extra data, so increasing the Taxa per Sample value beyond what was initially loaded (default top 10 per sample) doesn't actually do anything.

# Tests

* Verified that the Taxa per Sample slider works on the client-side when the `heatmap_filter_fe` flag is enabled.

### CODE REVIEW

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 1]
```diff
@@ -512,6 +511,53 @@ class SamplesHeatmapView extends React.Component {
...
+      allTaxonDetails,
+      selectedOptions,
+    } = this.state;
+    let { metric, taxonsPerSample } = selectedOptions;
+    let topTaxIds = new Set(),
+      topTaxonDetails = {},
+      filteredData = {};
+
+    Object.values(sampleDetails).forEach(sample => {
+      let filteredTaxaInSample = sample.taxa.filter(taxon =>
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: nit: can we call it `taxonId`?

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 2]
```diff
@@ -512,6 +511,53 @@ class SamplesHeatmapView extends React.Component {
...
+      for (let taxId of filteredTaxaInSample) {
+        if (count >= taxonsPerSample) {
+          break;
+        } else if (!topTaxIds.has(taxId)) {
+          let taxon = allTaxonDetails[taxId];
+          topTaxIds.add(taxId);
+          topTaxonDetails[taxId] = allTaxonDetails[taxId];
+          topTaxonDetails[taxon["name"]] = allTaxonDetails[taxId];
+
+          this.props.metrics.forEach(metric => {
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Please load props on top of the function just like you loaded state, i.e. `const { metrics } = this.props`

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 3]
```diff
@@ -512,6 +511,53 @@ class SamplesHeatmapView extends React.Component {
...
+    // Fetch the top N taxa from each sample, sorted by the selected metric,
+    // that passed all selected filters.
+    let {
+      sampleDetails,
+      allData,
+      allTaxonDetails,
+      selectedOptions,
+    } = this.state;
+    let { metric, taxonsPerSample } = selectedOptions;
+    let topTaxIds = new Set(),
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: nit: can you add extra line before?

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 4]
```diff
@@ -512,6 +511,53 @@ class SamplesHeatmapView extends React.Component {
...
+          break;
+        } else if (!topTaxIds.has(taxId)) {
+          let taxon = allTaxonDetails[taxId];
+          topTaxIds.add(taxId);
+          topTaxonDetails[taxId] = allTaxonDetails[taxId];
+          topTaxonDetails[taxon["name"]] = allTaxonDetails[taxId];
+
+          this.props.metrics.forEach(metric => {
+            filteredData[metric.value] = filteredData[metric.value] || [];
+            filteredData[metric.value][filteredData[metric.value].length] =
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Is this not just an array push?

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 5]
```diff
@@ -681,9 +727,9 @@ class SamplesHeatmapView extends React.Component {
         "thresholdFilters",
         "readSpecificity",
         "metric",
+        "taxonsPerSample",
       ];
-      // TODO(julie): taxaPerSample should eventually be a frontend filter
-      const backendFilters = ["background", "taxonsPerSample"];
+      const backendFilters = ["background"];
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: 💯

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx [hunk 6]
```diff
@@ -512,6 +511,53 @@ class SamplesHeatmapView extends React.Component {
...
+        } else if (!topTaxIds.has(taxId)) {
+          let taxon = allTaxonDetails[taxId];
+          topTaxIds.add(taxId);
+          topTaxonDetails[taxId] = allTaxonDetails[taxId];
+          topTaxonDetails[taxon["name"]] = allTaxonDetails[taxId];
+
+          this.props.metrics.forEach(metric => {
+            filteredData[metric.value] = filteredData[metric.value] || [];
+            filteredData[metric.value][filteredData[metric.value].length] =
+              allData[metric.value][taxon["index"]] || [];
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Just to confirm, this loads data for all samples for that particular tax id, right?

Comment (j-x-han) [reply app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Yes, I added a comment by `allData` as you suggested above that will hopefully make the data structure more clear!

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx (hunk 4 above)
Comment (j-x-han) [reply app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Oops, good catch.

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx (hunk 3 above)
Comment (j-x-han) [reply app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Done!

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx (hunk 2 above)
Comment (j-x-han) [reply app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Done!

File: app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx (hunk 1 above)
Comment (j-x-han) [reply app/assets/src/components/views/compare/SamplesHeatmapView/SamplesHeatmapView.jsx]: Done!

**Pull request merged**